presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java

```diff
@@ -186,6 +215,61 @@ public OrcReader(
         }
 
         this.cacheable = requireNonNull(cacheable, "hiveFileContext is null");
+
+        Optional<DwrfStripeCache> dwrfStripeCache = Optional.empty();
+        if (orcFileTail.getDwrfStripeCacheData().isPresent() && footer.getDwrfStripeCacheOffsets().isPresent()) {
+            DwrfStripeCache cache = buildDwrfStripeCache(footer.getStripes(), orcFileTail.getDwrfStripeCacheData().get(), footer.getDwrfStripeCacheOffsets().get());
+            dwrfStripeCache = Optional.of(cache);
+        }
+
+        requireNonNull(stripeMetadataSourceFactory, "stripeMetadataSourceFactory is null");
+        this.stripeMetadataSource = requireNonNull(stripeMetadataSourceFactory.create(dwrfStripeCache), "stripeMetadataSource is null");
+    }
+
+    private DwrfStripeCache buildDwrfStripeCache(List<StripeInformation> stripes, DwrfStripeCacheData dwrfStripeCacheData, List<Integer> cacheOffsets)
+    {
+        // AbstractOrcRecordReader sorts stripes by position, so probably there
+        // is a use-case when stripes can come unordered
+        List<StripeInformation> sortedStripes = new ArrayList<>(stripes);
+        sortedStripes.sort(comparingLong(StripeInformation::getOffset));
+        DwrfStripeCacheMode mode = dwrfStripeCacheData.getDwrfStripeCacheMode();
+        Slice cacheSlice = dwrfStripeCacheData.getDwrfStripeCacheSlice();
+
+        ImmutableMap.Builder<Long, Slice> indexSlices = ImmutableMap.builder();
+        if (mode.hasIndex()) {
+            for (int i = 0; i < sortedStripes.size(); i++) {
+                int offsetIdx = mode.getIndexOffsetPosition(i);
+                if (!addDwrfStripeCacheSlice(offsetIdx, cacheOffsets, cacheSlice, indexSlices, sortedStripes.get(i))) {
+                    break;
+                }
+            }
+        }
+
+        ImmutableMap.Builder<Long, Slice> footerSlices = ImmutableMap.builder();
+        if (mode.hasFooter()) {
+            for (int i = 0; i < sortedStripes.size(); i++) {
+                int offsetIdx = mode.getFooterOffsetPosition(i);
+                if (!addDwrfStripeCacheSlice(offsetIdx, cacheOffsets, cacheSlice, footerSlices, sortedStripes.get(i))) {
+                    break;
+                }
+            }
+        }
+
+        return new DwrfStripeCache(mode, indexSlices.build(), footerSlices.build());
+    }
+
+    private static boolean addDwrfStripeCacheSlice(int cacheOffsetIdx, List<Integer> cacheOffsets, Slice cacheSlice, ImmutableMap.Builder<Long, Slice> slices, StripeInformation stripe)
+    {
+        if (cacheOffsetIdx >= cacheOffsets.size() - 1) {
+            return false;
+        }
+
+        int sliceOffset = cacheOffsets.get(cacheOffsetIdx);
+        int sliceSize = cacheOffsets.get(cacheOffsetIdx + 1) - sliceOffset;
+        checkState(sliceOffset < cacheSlice.length());
```

Comment:
add, some debug information on what failed the checkState.

Reply:
done